src/JustSaying/Fluent/PublishConfig/DynamicPublicationConfiguration.cs

```diff
@@ -0,0 +1,97 @@
+using System.Collections.Concurrent;
+using System.ComponentModel;
+using System.Xml.Linq;
+using JustSaying.AwsTools;
+using JustSaying.AwsTools.QueueCreation;
+using JustSaying.Messaging;
+using JustSaying.Messaging.Interrogation;
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Fluent;
+
+internal class DynamicMessagePublisher : IMessagePublisher
+{
+    private readonly Dictionary<string, IMessagePublisher> _publisherCache = new();
+    private readonly Func<Message, string> _topicNameCustomizer;
+    private readonly Func<string, StaticPublicationConfiguration> _staticConfigBuilder;
+
+    private readonly ConcurrentDictionary<string, SemaphoreSlim> _topicCreationLocks = new();
+    private readonly ILogger<DynamicMessagePublisher> _logger;
+
+    public DynamicMessagePublisher(Func<Message, string> topicNameCustomizer, Func<string, StaticPublicationConfiguration> staticConfigBuilder, ILoggerFactory loggerFactory)
+    {
+        _topicNameCustomizer = topicNameCustomizer;
+        _staticConfigBuilder = staticConfigBuilder;
+        _logger = loggerFactory.CreateLogger<DynamicMessagePublisher>();
+    }
+
+    public InterrogationResult Interrogate()
+    {
+        return new InterrogationResult(new
+        {
+
+        });
+    }
+
+    public Task StartAsync(CancellationToken stoppingToken)
+    {
+        return Task.CompletedTask;
+    }
+
+    public async Task PublishAsync(Message message, PublishMetadata metadata, CancellationToken cancellationToken)
+    {
+        var topicName = _topicNameCustomizer(message);
+        if (_publisherCache.ContainsKey(topicName))
```

Comment:
Should this be `ConcurrentDictionary` too to prevent a race?

Reply:
Ah yeah fair - I thought it'd be ok as it'd only be concurrent readers on different keys, but other threads might be adding additional publishers so probably best to be on the safe side 👍 .